File: LeetCode-Java/Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        ArrayList<Integer> num = new ArrayList<Integer>();
        while(head != null)
        {
            num.add(head.val);
            head = head.next;
        }
        return arrayToBST(num, 0, num.size() - 1);
    }
    
    TreeNode arrayToBST(ArrayList<Integer> num, int start, int end)
    {
        if(end < start)
            return null;
        if(end == start)
            return new TreeNode(num.get(start));
        else
        {
            int mid = (start + end) / 2;
            TreeNode tempNode = new TreeNode(num.get(mid));
            tempNode.left = arrayToBST(num, start, mid - 1);
            tempNode.right = arrayToBST(num, mid + 1, end);
            return tempNode;
        }
    }  
}

File: Leetcode-cpp/convert_sorted_list_to_binary_search_tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (head == nullptr) return nullptr;
        if (head->next == nullptr) return new TreeNode(head->val);
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast->next != nullptr && fast->next->next != nullptr) {
            fast = fast->next->next;
            if (fast->next == nullptr) break;
            slow = slow->next;
        }
        TreeNode* root = new TreeNode(slow->next->val);
        root->right = sortedListToBST(slow->next->next);
        slow->next = nullptr;
        root->left = sortedListToBST(head);
        return root;
    }
};